lock-api/src/main/java/com/palantir/lock/client/UnreliableTimeLockService.java

```diff
@@ -31,30 +32,52 @@
 import com.palantir.logsafe.SafeArg;
 import com.palantir.logsafe.logger.SafeLogger;
 import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.timestamp.TimestampManagementService;
 import com.palantir.timestamp.TimestampRange;
+import java.security.SecureRandom;
 import java.util.List;
 import java.util.Set;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
 import java.util.stream.Collectors;
 
 /**
  * A version of timelock service that has a chance to randomly lose locks during refresh or immediately after
  * acquiring. This is useful for testing the behavior of clients when locks are lost.
  */
 public final class UnreliableTimeLockService implements TimelockService {
+    // A partition is sized at 2^23 timestamps, and we want to skip through potentially many at a time.
+    private static final long THOUSAND_COARSE_PARTITIONS_SIZE = 1000 * (1L << 23);
+    private static final SecureRandom SECURE_RANDOM = DefaultNativeSamplingSecureRandomFactory.INSTANCE.create();
 
     private static final SafeLogger log = SafeLoggerFactory.get(UnreliableTimeLockService.class);
 
     private final TimelockService delegate;
+    private final TimestampManagementService managementService;
     private final BuggifyFactory buggify;
+    private final ReadWriteLock timestampLock;
 
-    public static UnreliableTimeLockService create(TimelockService timelockService) {
-        return new UnreliableTimeLockService(timelockService, DefaultBuggifyFactory.INSTANCE);
+    public static UnreliableTimeLockService create(
+            TimelockService timelockService, TimestampManagementService managementService) {
+        return new UnreliableTimeLockService(timelockService, managementService, DefaultBuggifyFactory.INSTANCE);
     }
 
     @VisibleForTesting
-    UnreliableTimeLockService(TimelockService delegate, BuggifyFactory buggifyFactory) {
+    UnreliableTimeLockService(
+            TimelockService delegate,
+            TimestampManagementService timeLockManagementService,
+            BuggifyFactory buggifyFactory) {
         this.delegate = delegate;
+        this.managementService = timeLockManagementService;
         this.buggify = buggifyFactory;
+
+        // Fast forward doesn't actually verify it's going forward, so we need to make sure we don't end up
+        // with a situation where we buggify -> getFreshTimestamp -> calculate next timestamp -> get stuck
+        // while the timestamp progresses, then the original thread then fast forwards (but actually goes back in time)
+        // It's an unfair lock to add more chaos to the buggified requests.
```

Comment:
We spoke offline about fair/unfair, determinism and other things here - would be amazing if there was some antithesis provided thing for locking. we should keep unfair for now